mlrun/utils/retryer.py

```diff
@@ -0,0 +1,206 @@
+        yield next_value
+        x += 1
+
+
+def create_step_backoff(steps=None):
+    """
+    Create a generator of steps backoff.
+    Example: steps = [[2, 5], [20, 10], [120, None]] will produce a generator in which the first 5
+    values will be 2, the next 10 values will be 20 and the rest will be 120.
+    :param steps: a list of lists [step_value, number_of_iteration_in_this_step]
+    """
+    steps = steps if steps is not None else [[2, 10], [10, 10], [120, None]]
+    steps = iter(steps)
+
+    # Get first step
+    step = next(steps)
+    while True:
+        current_step_value, current_step_remain = step
+        if current_step_remain == 0:
+            # No more in this step, moving on
+            step = next(steps)
+        elif current_step_remain is None:
+            # We are in the last step, staying here forever
+            yield current_step_value
+        elif current_step_remain > 0:
+            # Still more remains in this step, just reduce the remaining number
+            step[1] -= 1
+            yield current_step_value
+
+
+def create_exponential_backoff(base=2, max_value=120, scale_factor=1):
+    """
+    Create a generator of exponential backoff. Check out usage example in test_helpers.py
+    :param base: exponent base
+    :param max_value: max limit on the result
+    :param scale_factor: factor to be used as linear scaling coefficient
+    """
+    exponent = 1
+    while True:
+        # This "complex" implementation (unlike the one in linear backoff) is to avoid exponent growing too fast and
+        # risking going behind max_int
+        next_value = scale_factor * (base**exponent)
+        if next_value < max_value:
+            exponent += 1
+            yield next_value
+        else:
+            yield max_value
+
+
+class Retryer:
+    def __init__(self, backoff, timeout, logger, verbose, function, *args, **kwargs):
+        """
+        Initialize function retryer with given *args and **kwargs.
+        Tries to run it until success or timeout reached (timeout is optional)
+        :param backoff: can either be a:
+                - number (int / float) that will be used as interval.
+                - generator of waiting intervals. (support next())
+        :param timeout: pass None if timeout is not wanted, number of seconds if it is
+        :param logger: a logger so we can log the failures
+        :param verbose: whether to log the failure on each retry
+        :param _function: function to run
+        :param args: functions args
+        :param kwargs: functions kwargs
+        """
+        self.backoff = backoff
+        self.timeout = timeout
+        self.logger = logger
+        self.verbose = verbose
+        self.function = function
+        self.args = args
+        self.kwargs = kwargs
+        self.start_time = None
+        self.last_exception = None
+        self.first_interval = None
+        self._prepare()
```

Comment:
why not under `run()`? better fitted there and start time will be more accurate?